0000-stream/0000-stream.md

```diff
@@ -0,0 +1,405 @@
+
+#### 4.4.5. Closing Streams
+
+Either endpoint can close a stream using a `StreamClose` frame. Implementations MAY allow half-open streams.
+
+`StreamClose` frames are used to communicate both normal stream closes as well as errors.
+
+### 4.5. Connection-Level Flow Control
+
+Similar to the [stream-level flow control](#4-4-4-stream-level-flow-control), endpoints can limit the total amount of incoming data on all streams. Endpoints advertise the total number of bytes they are willing to receive on a given connection using `ConnectionMaxData` frames. Endpoints MAY increase the advertised limits by sending additional `ConnectionMaxData` frames with higher total byte limits.
+
+The receiving endpoint MUST close the connection with a `FlowControlError` if the sender violates the advertised limit.
+
+### 4.6. Closing Connections
+
+Either endpoint can close the connection using a `ConnectionClose` frame. Implementations MAY allow half-open connections.
+
+`ConnectionClose` frames are used to communicate both normal connection closes as well as errors.
+
+## 5. Packet and Frame Specification
+
+STREAM packets are encrypted and attached to ILP Prepare, Fulfill, and Reject packets as the `data` field.
+
+STREAM uses the [Octet Encoding Rules (OER)](http://www.oss.com/asn1/resources/books-whitepapers-pubs/Overview_of_OER.pdf) to encode packet fields.
+
+### 5.1. Encryption
+
+All STREAM packets are encrypted using [AES-256-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a 12-byte Initialization Vector (IV) and a 16-Byte Authentication Tag.
+
+If subsequent versions support additional encryption algorithms, those details should be exchanged between the sender and receiver when they establish the shared secret. If a receiver attempts to decrypt an incoming packet but is unable to (perhaps because the sender is using an unsupported cipher), the receiver SHOULD reject the incoming transfer with an `F06: Unexpected Payment` error.
+
+#### 5.1.1. Encryption Envelope
+
+See the [ASN.1 definition](../asn1/Stream.asn) for the formal encryption envelope specification.
+
+| Field | Type | Description |
+|---|---|---|
+| Random IV  | 12-Byte UInt | Nonce used as input to the AES-GCM algorithm. Also ensures conditions are random. Endpoints MUST NOT encrypt two packets with the same nonce |
+| Authentication Tag | 16-Byte UInt | Authentication tag produced by AES-GCM encryption that ensures data integrity |
+| Ciphertext | 0-32739 Bytes | Encrypted data (see below for contents) |
+
+Note that the `Ciphertext` is NOT length-prefixed. The length can be inferred from the ILP packet because the whole `data` field is encoded with a length prefix. (This is done so that the entire `data` field is indistinguishable from random bytes.)
+
+#### 5.1.2. Encryption Pseudocode
+
+The encryption key used for every packet sent for a given connection is derived from the shared secret and the string `"ilp_stream_encryption"` using an HMAC.
```

Comment:
Since it's all plain ascii characters, the string is the same in ASCII and UTF8

Reply:
True. Although, I think specifying the encoding is still worth doing because crypto algorithms are generally defined with either numbers or octet streams as input not strings. Specifying the encoding just explicitly tells the reader how to go from the string to the desired octets.